examples/sndcardparser/sample_sndcardparser.c

```diff
@@ -0,0 +1,269 @@
+
+enum snd_node_type {
+    NODE_TYPE_HW = 0,
+    NODE_TYPE_PLUGIN,
+    NODE_TYPE_INVALID,
+};
+
+enum {
+    NODE_PCM,
+    NODE_COMPR,
+    NODE_MIXER,
+    NODE_MAX,
+};
+
+struct snd_node_ops {
+    /** Function pointer to get card definition */
+    void* (*open_card)(unsigned int card);
+    /** Function pointer to release card definition */
+    void (*close_card)(void *card);
+    /** Get interger type properties from device definition */
+    int (*get_int)(void *node, const char *prop, int *val);
+    /** Get string type properties from device definition */
+    int (*get_str)(void *node, const char *prop, char **val);
+    /** Function pointer to get mixer definition */
+    void* (*get_mixer)(void *card);
+    /** Function pointer to get PCM definition */
+    void* (*get_pcm)(void *card, unsigned int id);
+    /** Function pointer to get COMPRESS definition */
+    void* (*get_compress)(void *card, unsigned int id);
+};
+
+struct snd_dev_def {
+    unsigned int device;
+    int type;
+    const char *name;
+    const char *so_name;
+    int playback; //used only for pcm node
+    int capture;  //used only for pcm node
+    /* add custom props here */
+};
+
+struct snd_dev_def_card {
+    unsigned int card;
+    char *name;
+
+    /* child device details */
+    int num_pcm_nodes;
+    struct snd_dev_def *pcm_dev_def;
+
+    struct snd_dev_def *mixer_dev_def;
+};
+
+struct snd_dev_def pcm_devs[] = {
+    {100, NODE_TYPE_PLUGIN, "PCM100", "libsample_pcm_plugin.so", 1, 0},
+    /* Add other plugin info here */
+};
+
+struct snd_dev_def mixer_dev =
+    {VIRTUAL_SND_CARD_ID, NODE_TYPE_PLUGIN, "virtual-snd-card", "libsample_mixer_plugin.so", 0, 0};
+
+void *snd_card_def_get_card(unsigned int card)
+{
+    struct snd_dev_def_card *card_def = NULL;
+    struct snd_dev_def *pcm_dev_def = NULL;
+    struct snd_dev_def *mixer_dev_def = NULL;
+    int num_pcm = ARRAY_SIZE(pcm_devs);
+    int i;
+
+    if (card != VIRTUAL_SND_CARD_ID)
+        return NULL;
+
+    card_def = calloc(1, sizeof(struct snd_dev_def_card));
+    if (!card_def)
+        return card_def;
+
+    card_def->card = card;
+    card_def->name = strdup("virtual-snd-card");
+
+    /* fill pcm device node info */
+    card_def->num_pcm_nodes = num_pcm;
+    pcm_dev_def = calloc(num_pcm, sizeof(struct snd_dev_def));
+    if (!pcm_dev_def)
+        goto free_card_def;
+
+    for (i = 0; i < num_pcm; i++)
+        memcpy(&pcm_dev_def[i], &pcm_devs[i], sizeof(struct snd_dev_def));
+
+    card_def->pcm_dev_def = pcm_dev_def;
+
+    /* fill mixer device node info */
+    mixer_dev_def = calloc(1, sizeof(struct snd_dev_def));
+    if (!mixer_dev_def)
+        goto free_pcm_dev;
+
+    memcpy(mixer_dev_def, &mixer_dev, sizeof(struct snd_dev_def));
+
+    card_def->mixer_dev_def = mixer_dev_def;
+    return card_def;
+free_pcm_dev:
+    free(pcm_dev_def);
+free_card_def:
+    free(card_def->name);
+    free(card_def);
+    return NULL;
+}
+
+void snd_card_def_put_card(void *card_node)
```

Comment:
This should get renamed to `snd_card_def_close_card`.

Reply:
Agree. It would also be good to rename the `snd_card_def_get_card` to `snd_card_def_open_card`.

Reply:
sure .. will update the patch